pkg/volume/volume_linux_test.go

```diff
@@ -158,5 +163,176 @@ func TestSkipPermissionChange(t *testing.T) {
 
 		})
 	}
+}
+
+func TestSetVolumeOwnership(t *testing.T) {
+	always := v1.AlwaysChangeVolumePermission
+	onrootMismatch := v1.OnRootMismatch
+	expectedMask := rwMask | os.ModeSetgid | execMask
+
+	tests := []struct {
+		description         string
+		fsGroupChangePolicy *v1.PodFSGroupChangePolicy
+		setupFunc           func(path string) error
+		assertFunc          func(path string) error
+		featureGate         bool
+	}{
+		{
+			description:         "featuregate=on, fsgroupchangepolicy=always",
+			fsGroupChangePolicy: &always,
+			featureGate:         true,
+			setupFunc: func(path string) error {
+				info, err := os.Lstat(path)
+				if err != nil {
+					return err
+				}
+				// change mode of root folder to be right
+				err = os.Chmod(path, info.Mode()|expectedMask)
+				if err != nil {
+					return err
+				}
+
+				// create a subdirectory with invalid permissions
+				rogueDir := filepath.Join(path, "roguedir")
+				err = os.Mkdir(rogueDir, info.Mode())
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			assertFunc: func(path string) error {
+				rogueDir := filepath.Join(path, "roguedir")
+				info, err := os.Lstat(path)
+				if err != nil {
+					return err
+				}
+				stat, ok := info.Sys().(*syscall.Stat_t)
+				if !ok || stat == nil {
+					return fmt.Errorf("unable to get stat information for %s", rogueDir)
+				}
+
+				var expectedGid int64 = int64(stat.Gid)
+				doPermissionChange := requiresPermissionChange(rogueDir, &expectedGid, false)
```

Comment:
Can we do a simpler check to just verify the permission bits? Depending on this function to assert in the unit test means that if there's bug in that function, the unit test may not catch it.

Reply:
fixed